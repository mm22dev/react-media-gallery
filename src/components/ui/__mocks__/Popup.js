const Popup = () => <div className='popup-mock'></div>

Popup.displayName = 'PopupMock'

export default Popup
