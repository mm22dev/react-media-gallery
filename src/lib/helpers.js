const numbConv = cssProp => Number.isNaN(parseFloat(cssProp)) ? 0 : parseFloat(cssProp)

const isNodeEl = el => el !== null && typeof el === 'object' && 'nodeType' in el

export const S = (selector, ancestorEl = null) =>
  isNodeEl(ancestorEl)
    ? ancestorEl.querySelectorAll(selector).length > 0
      ? ancestorEl.querySelectorAll(selector)
      : null
    : document.querySelectorAll(selector).length > 0
      ? document.querySelectorAll(selector)
      : null

export const gcs = el => window.getComputedStyle(el)

export const cloneObj = obj => JSON.parse(JSON.stringify(obj))

export const getHeight = el => {
  if (el !== null && isNodeEl(el)) {
    const height = el.offsetHeight
    const borderTopWidth = gcs(el).getPropertyValue('border-top-width')
    const borderBottomWidth = gcs(el).getPropertyValue('border-bottom-width')
    const paddingTop = gcs(el).getPropertyValue('padding-top')
    const paddingBottom = gcs(el).getPropertyValue('padding-bottom')
    return height - numbConv(borderBottomWidth) - numbConv(borderTopWidth) - numbConv(paddingTop) - numbConv(paddingBottom)
  } else {
    return 0
  }
}

export const getWidth = el => {
  if (el !== null && isNodeEl(el)) {
    const width = el.offsetWidth
    const borderLeftWidth = gcs(el).getPropertyValue('border-left-width')
    const borderRightWidth = gcs(el).getPropertyValue('border-left-width')
    const paddingLeft = gcs(el).getPropertyValue('padding-left')
    const paddingRight = gcs(el).getPropertyValue('padding-right')
    return width - numbConv(borderRightWidth) - numbConv(borderLeftWidth) - numbConv(paddingLeft) - numbConv(paddingRight)
  } else {
    return 0
  }
}

export const identityFn = f => f

export const emptyQueueObject = {
  src: null,
  captionText: null,
  width: null,
  height: null,
  type: null,
  object: null,
  options: null
}

export const emptyMediaContent = {
  tagFamily: null,
  innerProps: {}
}

export const emptySpringSettings = {
  from: {},
  to: {},
  config: {},
  immediate: false,
  onRest: null
}

export const emptyActionReducer = {
  newState: null,
  callback: null
}

export const pixelify = size => `${size}px`

export const depixelify = size => +size.replace(/[^\d.-]/g, '')

export const hide = elementsArray => elementsArray.map(el => { el.style = { ...el.style, display: 'none' } })

export const show = (elementsArray, displayVariant = null) => elementsArray.map(el => { el.style = { ...el.style, display: (displayVariant !== null) ? displayVariant : 'block' } })

export const fadeOut = (elementsArray, duration = 400) => elementsArray.map(el => { el.fade = duration === 0 ? { inProp: false, immediate: true } : { inProp: false, duration: duration, immediate: false } })

export const fadeIn = (elementsArray, duration = 400) => elementsArray.map(el => { el.fade = duration === 0 ? { inProp: true, immediate: true } : { inProp: true, duration: duration, immediate: false } })

export class PRTError extends Error {
  constructor (message) {
    super(message)
    this.name = 'PopReactrox Warning'
  }
}
