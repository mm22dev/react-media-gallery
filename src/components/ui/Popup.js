import React, { Component, createRef } from 'react'
import PropTypes from 'prop-types'
import { connectWithPopupSettings } from '../../providers/PopupSettings'
import initData from '../../data/popupDefaultData.json'
import C from '../../actions/constants'
import getAction from '../../actions/index'
import Loader from './Loader'
import Media from './Media'
import Caption from './Caption'
import Closer from './Closer'
import Navbar from './Navbar'
import { pixelify, identityFn, S, getHeight, PRTError } from '../../lib/helpers'
import isEqual from 'lodash.isequal'
import { Spring } from 'react-spring/renderprops'

class PopupUI extends Component {
  constructor (props) {
    super(props)
    this.self = createRef()
    this.popupSettings = this.props.popupSettings
    const { usePopupDefaultStyling, popupBackgroundColor, popupTextColor, popupPadding, popupCaptionHeight, usePopupCaption, popupWidth, popupHeight } = this.popupSettings
    this.style = {
      display: 'none',
      verticalAlign: 'middle',
      position: 'relative',
      zIndex: 1,
      cursor: 'pointer',
      ...(usePopupDefaultStyling && {
        background: popupBackgroundColor,
        color: popupTextColor,
        padding: pixelify(popupPadding)
      }),
      ...(usePopupDefaultStyling && usePopupCaption && { paddingBottom: pixelify(popupCaptionHeight) }),
      minWidth: pixelify(popupWidth),
      minHeight: pixelify(popupHeight)
    }
    this.state = {
      springAnimationSettings: initData.springAnimationSettings.default,
      renderSettings: initData.renderSettings,
      structureSettings: initData.structureSettings
    }
    this.handleErrors = this.handleErrors.bind(this)
    this.performAction = this.performAction.bind(this)
    this.getLoaderContentMarginTop = this.getLoaderContentMarginTop.bind(this)
    this.getSpringAnimationSettings = this.getSpringAnimationSettings.bind(this)
    this.onWindowResize = () => this.performAction(C.ON_WINDOW_RESIZE_REQ.req)
    this.onKeyDown = ev => this.performAction(C.ON_KEY_DOWN.req, ev)
  }

  handleErrors (errors) {
    try {
      if (errors.length > 0) { throw new PRTError(errors[0]) }
    } catch (err) {
      errors.shift()
      this.setState(prevState => ({
        structureSettings: {
          ...prevState.structureSettings,
          errors
        }
      }), console.error(err))
    }
  }

  performAction (action, ...args) {
    let { newState, callback } = getAction(action)(this, ...args)
    callback = callback !== null ? callback : identityFn

    newState !== null
      ? this.setState(prevState => ({
        ...prevState,
        ...newState
      }), callback)
      : callback()
  }

  getLoaderContentMarginTop () {
    const DOMref = this.self.current
    const { popupCaptionSelector, popupLoaderSelector } = this.popupSettings
    const computedPopupHeight = getHeight(DOMref)
    const computedCaptionHeight = getHeight(
      S(popupCaptionSelector, DOMref) !== null
        ? S(popupCaptionSelector, DOMref)[0]
        : null
    )
    const computedLoaderContentHeight = getHeight(
      S(`${popupLoaderSelector} div`, DOMref) !== null
        ? S(`${popupLoaderSelector} div`, DOMref)[0]
        : null
    )
    return Math.floor((computedPopupHeight - computedLoaderContentHeight + computedCaptionHeight) / 2)
  }

  getSpringAnimationSettings () {
    const { triggerAnimation, duration, immediate, sizes } = this.state.springAnimationSettings
    const mediaContentStatus = this.state.renderSettings.popup.childs.media.content.status

    return {
      from: triggerAnimation ? this.state.renderSettings.popup.prevSizes : sizes,
      to: sizes,
      config: { duration: duration },
      immediate: immediate,
      onRest: () => triggerAnimation && mediaContentStatus !== 'resize'
        ? this.performAction(C.ON_POPUP_SIZES_ANIMATION_END.req, initData.springAnimationSettings.default, sizes)
        : mediaContentStatus === 'loaded'
          ? this.performAction(C.ON_MEDIA_CONTENT_APPEAR_REQ.req)
          : identityFn
    }
  }

  shouldComponentUpdate (nextProps, nextState) {
    return !isEqual(this.props, nextProps) || !isEqual(this.state, nextState)
  }

  componentDidUpdate (prevProps) {
    const hasOverlayStateChanged = !isEqual(prevProps.overlayState, this.props.overlayState)
    const overlayStatusList = {
      opening: () => this.state.structureSettings.queue[this.props.overlayState.anchorIndex] !== null
        ? this.performAction(C.ON_PRT_OPEN.req)
        : this.props.updateOverlayStateWith({ status: 'closed' }),
      opened: () => this.performAction(C.ON_PRT_SWITCH.req, this.props.overlayState.anchorIndex, true),
      closing: () => this.performAction(C.ON_PRT_CLOSE.req),
      closed: () => this.performAction(C.ON_OVERLAY_FADE_OUT_END.req),
      refreshing: () => this.performAction(C.ON_QUEUE_GENERATION.req, this.props.updateOverlayStateWith, true),
      refreshed: () => this.props.updateOverlayStateWith({
        status: 'opening',
        anchorIndex: this.props.getClickedAnchorIndex(this.props.overlayState.pendingMedia),
        pendingMedia: null
      })
    }
    const callbackOnUpdate = (this.props.overlayState.status in overlayStatusList) && hasOverlayStateChanged
      ? overlayStatusList[this.props.overlayState.status]
      : identityFn
    callbackOnUpdate()
    this.handleErrors([...this.state.structureSettings.errors])
  }

  render () {
    const { popup } = this.state.renderSettings
    const { childs } = popup
    const popupStyle = { ...this.style, ...popup.style }
    const springAnimationSettings = { ...this.getSpringAnimationSettings() }
    const { animate } = this.state.renderSettings.popup.childs.loader

    return (
      <Spring {...springAnimationSettings}>
        {
          springProps => (
            <div className={`prt-p-default ${this.popupSettings.popupClass}`} style={{ ...popupStyle, ...springProps }} onClick={ev => this.performAction(C.ON_POPUP_CLICK.req, ev)} ref={this.self} >
              <Loader {...childs.loader} {...(animate && { innerMarginTop: this.getLoaderContentMarginTop() })} />
              <Media {...childs.media} structureSettings={this.state.structureSettings} actionPerformer={this.performAction} />
              <Caption {...childs.caption} />
              <Closer {...childs.closer} updateOverlayStateWith={this.props.updateOverlayStateWith} />
              <Navbar {...childs.navbar} actionPerformer={this.performAction} />
            </div>
          )
        }
      </Spring>
    )
  }

  componentDidMount () {
    this.performAction(C.ON_QUEUE_GENERATION.req, newOverlayState => this.props.updateOverlayStateWith(newOverlayState), false)
    window.addEventListener('resize', this.onWindowResize)
    window.addEventListener('keydown', this.onKeyDown)
  }

  componentWillUnmount () {
    window.removeEventListener('resize', this.onWindowResize)
    window.removeEventListener('keydown', this.onKeyDown)
  }
}

PopupUI.propTypes = {
  popupSettings: PropTypes.object,
  overlayState: PropTypes.object,
  getClickedAnchorIndex: PropTypes.func,
  updateOverlayStateWith: PropTypes.func
}

PopupUI.defaultProps = {
  overlayState: {
    anchorIndex: null,
    status: 'init'
  },
  getClickedAnchorIndex: identityFn,
  updateOverlayStateWith: identityFn
}

const Popup = connectWithPopupSettings(PopupUI)
Popup.displayName = 'Popup'
export default Popup
