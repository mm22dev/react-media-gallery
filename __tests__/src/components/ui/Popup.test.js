import Popup from '../../../../src/components/ui/Popup'
import toJSON from 'enzyme-to-json'
import { compose } from 'redux'
import defaultPopupState from '../../../../src/data/popupDefaultData.json'
import defaultPopupSettings from '../../../../src/data/popupSettings.json'
import { cloneObj } from '../../../../src/lib/helpers'

jest.mock('../../../../src/components/ui/Caption')
jest.mock('../../../../src/components/ui/Closer')
jest.mock('../../../../src/components/ui/Loader')
jest.mock('../../../../src/components/ui/Media')
jest.mock('../../../../src/components/ui/Navbar')

const { shallow, mount } = Enzyme

describe('<Popup /> Component', () => {
  const { popupSettings } = defaultPopupSettings
  let wrapper
  let _performAction = jest.fn()

  describe('Rendering UI.', () => {
    it('Renders correctly.', () => {
      compose(expect, toJSON, shallow)(<Popup />).toMatchSnapshot()
    })

    it('Click does not cause Error.', () => {
      wrapper = mount(<Popup popupSettings={popupSettings} />)
      _performAction = jest.fn()
      wrapper.find('PopupUI').instance().performAction = _performAction
      wrapper.find('.prt-p-default').simulate('click')
      expect(_performAction.mock.calls[0][0]).toBe('onPopupClick')
    })

    it('Further bottom padding is added when captions are enabled.', () => {
      wrapper = mount(<Popup popupSettings={{ ...popupSettings, usePopupDefaultStyling: true, usePopupCaption: true }} />)
      expect(wrapper.find('PopupUI').instance().style.paddingBottom).toBe(`${popupSettings.popupCaptionHeight}px`)
    })

    describe('Performing actions.', () => {
      let actionPerformer
      let _mockedState = defaultPopupState
      let _prevState
      let _newState

      beforeAll(() => {
        wrapper = mount(<Popup popupSettings={{ ...popupSettings, overlayClass: 'mock-class' }} />)
        _mockedState.structureSettings.queue = ['m', 'o', 'c', 'k']
        wrapper.find('PopupUI').instance().setState({ structureSettings: _mockedState.structureSettings })
        _prevState = wrapper.find('PopupUI').instance().state
        actionPerformer = wrapper.find('PopupUI').instance().performAction
      })

      it('Does not throw Error when nonnexistent action is provided.', () => {
        actionPerformer('mock-action')
        _newState = wrapper.find('PopupUI').instance().state
        expect(cloneObj(_prevState)).toEqual(cloneObj(_newState))
      })

      it('Updates the state and invokes callback at the end.', () => {
        actionPerformer('onPRTOpen')
        _newState = wrapper.find('PopupUI').instance().state
        expect(cloneObj(_prevState)).not.toEqual(cloneObj(_newState))
        expect(_newState.structureSettings.isLocked).toBe(true)
      })
    })

    describe('Requesting loader animation.', () => {
      const _getLoaderContentMarginTop = jest.fn()
      const _popupSettings = {
        ...popupSettings,
        popupPadding: 0,
        usePopupCaption: true,
        popupCaptionHeight: 0
      }
      let _popupState

      it('innerMarginTop prop is passed to <Loader />', () => {
        wrapper = mount(<Popup popupSettings={_popupSettings} />)
        wrapper.find('PopupUI').instance().getLoaderContentMarginTop = _getLoaderContentMarginTop
        _popupState = cloneObj(defaultPopupState)
        _popupState.renderSettings.popup.childs.loader.animate = true
        wrapper.find('PopupUI').instance().setState({ ..._popupState })
        expect(_getLoaderContentMarginTop).toHaveBeenCalled()
      })

      it('innerMarginTop is conditioned by the heights of <Loader /> and <Caption />', () => {
        wrapper = mount(<Popup popupSettings={_popupSettings} />)
        expect(wrapper.find('PopupUI').instance().getLoaderContentMarginTop()).toBe(0)
      })

      it('innerMarginTop depends only on <Popup /> sizes.', () => {
        wrapper = mount(<Popup popupSettings={{ ..._popupSettings, popupCaptionSelector: '.wrong-caption-selector', popupLoaderSelector: '.wrong-loader-selector' }} />)
        expect(wrapper.find('PopupUI').instance().getLoaderContentMarginTop()).toBe(0)
      })
    })

    describe('Assigning props to <Spring />', () => {
      let _popupState,
        _performAction,
        _springAnimationSettings

      beforeAll(() => {
        _popupState = cloneObj(defaultPopupState)
        _performAction = jest.fn()
        wrapper = mount(<Popup popupSettings={popupSettings} />)
      })

      beforeEach(() => {
        _springAnimationSettings = {
          triggerAnimation: false,
          duration: 300,
          immediate: false,
          sizes: {
            width: '100px',
            height: '100px'
          }
        }
      })

      it("Popup's resizer method is invoked when required.", done => {
        _springAnimationSettings.triggerAnimation = true
        wrapper.find('PopupUI').instance().performAction = jest.fn(args => {
          _performAction(args)
          expect(_performAction.mock.calls[0][0]).toBe('onPopupSizesAnimationEnd')
          done()
        })
        wrapper.find('PopupUI').setState({ springAnimationSettings: _springAnimationSettings })
      })

      it('<Media /> content is shown when it is loaded.', done => {
        _popupState.renderSettings.popup.childs.media.content.status = 'loaded'
        wrapper.find('PopupUI').instance().performAction = jest.fn(args => {
          _performAction(args)
          expect(_performAction).toBeCalledWith('onMediaContentAppearReq')
          done()
        })
        wrapper.find('PopupUI').setState({ renderSettings: _popupState.renderSettings, springAnimationSettings: _springAnimationSettings })
      })
    })

    describe('Receiving props from parent component', () => {
      const _overlayState = {
        status: 'opening',
        anchorIndex: 0
      }

      beforeAll(() => {
        wrapper = mount(<Popup popupSettings={popupSettings} />)
        wrapper.find('PopupUI').instance().performAction = _performAction
      })

      it('Invokes the method to lock popup modal.', () => {
        wrapper.setProps({ overlayState: _overlayState })
        expect(_performAction.mock.calls[0][0]).toBe('onPRTOpen')
      })

      it('Invokes the method to start the media loading process.', () => {
        wrapper.setProps({ overlayState: { ..._overlayState, status: 'opened' } })
        expect(_performAction.mock.calls[0][0]).toBe('onPRTSwitch')
        expect(_performAction.mock.calls[0][1]).toBe(_overlayState.anchorIndex)
      })

      it('Invokes the method to close popup modal.', () => {
        wrapper.setProps({ overlayState: { ..._overlayState, status: 'closing' } })
        expect(_performAction.mock.calls[0][0]).toBe('onPRTClose')
      })

      it('Invokes the method to unlock popup modal.', () => {
        wrapper.setProps({ overlayState: { ..._overlayState, status: 'closed' } })
        expect(_performAction.mock.calls[0][0]).toBe('onOverlayFadeOutEnd')
      })

      it('Excludes the media from being shown.', () => {
        wrapper.find('PopupUI').instance().setState({
          structureSettings: {
            ...wrapper.find('PopupUI').instance().state.structureSettings,
            queue: [null]
          }
        })
        wrapper.setProps({ overlayState: _overlayState })
        expect(_performAction).not.toHaveBeenCalled()
      })

      it('Forces media queue to be regenerated.', () => {
        wrapper.setProps({ overlayState: { ..._overlayState, status: 'refreshing' } })
        expect(_performAction.mock.calls[0][0]).toBe('onQueueGeneration')
      })

      it('Forces the popup opening after media queue regeneration.', () => {
        const _updateOverlayStateWith = jest.fn()
        wrapper = mount(<Popup popupSettings={popupSettings} updateOverlayStateWith={_updateOverlayStateWith} />)
        wrapper.setProps({ overlayState: { ..._overlayState, status: 'refreshed' } })
        // updateOverlayStateWith first time is auto called by onQueueGeneration
        expect(_updateOverlayStateWith.mock.calls[1][0].status).toBe('opening')
        expect(_updateOverlayStateWith.mock.calls[1][0]).toHaveProperty('anchorIndex')
        expect(_updateOverlayStateWith.mock.calls[1][0].pendingMedia).toBe(null)
      })
    })

    describe('Mounting <Popup />', () => {
      const map = {}

      beforeAll(() => {
        window.addEventListener = jest.fn((event, cb) => {
          map[event] = cb
        })
        wrapper = mount(<Popup popupSettings={popupSettings} />)
        wrapper.find('PopupUI').instance().performAction = _performAction
      })

      it('Resize listener is added to window', () => {
        map.resize()
        expect(_performAction).toHaveBeenCalledWith('onWindowResizeReq')
      })

      it('KeyDown listener is added to window', () => {
        const _key = { key: 'Escape' }
        map.keydown(_key)
        expect(_performAction.mock.calls[0][0]).toBe('onKeyDown')
        expect(_performAction.mock.calls[0][1]).toBe(_key)
      })
    })
  })

  describe('Unomunting <Popup />', () => {
    const map = {}

    beforeAll(() => {
      window.addEventListener = jest.fn((event, cb) => {
        map[event] = cb
      })
      window.removeEventListener = jest.fn((event) => {
        delete map[event]
      })
    })

    beforeEach(() => {
      wrapper = mount(<Popup popupSettings={popupSettings} />)
    })

    it('The resize listener is appropriately removed.', () => {
      wrapper.find('PopupUI').instance().componentWillUnmount()
      expect(map).not.toHaveProperty('resize')
    })

    it('The keydown listener is appropriately removed.', () => {
      wrapper.find('PopupUI').instance().componentWillUnmount()
      expect(map).not.toHaveProperty('keydown')
    })
  })

  describe('Handling Errors.', () => {
    let _prevState

    beforeAll(() => {
      wrapper = mount(<Popup popupSettings={popupSettings} />)
      _prevState = wrapper.find('PopupUI').instance().state
    })

    it('If present, errors are thrown', () => {
      wrapper.find('PopupUI').instance().setState({
        structureSettings: {
          ..._prevState.structureSettings,
          errors: ['mock-error']
        }
      })
      expect(() => { wrapper.find('PopupUI').instance().handleErrors() }).toThrowError()
      expect(wrapper.find('PopupUI').instance().state.structureSettings.errors.length).toBe(0)
    })
  })
})
