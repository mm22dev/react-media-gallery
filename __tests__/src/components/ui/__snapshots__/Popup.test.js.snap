// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> Component Rendering UI. Renders correctly. 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
